device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttMessaging.java

```diff
@@ -154,15 +154,15 @@ private boolean appendPropertyIfPresent(StringBuilder stringBuilder, boolean sep
 
                 if (isApplicationProperty)
                 {
-                    stringBuilder.append(URLEncoder.encode(propertyKey, StandardCharsets.UTF_8.name()));
+                    stringBuilder.append(URLEncoder.encode(propertyKey, StandardCharsets.UTF_8.name()).replaceAll("\\+", "%20"));
```

Comment:
What's the limitation with `URLEncoder` being unable to encode `+`?

Reply:
https://www.urlencoder.io/java/

Reply:
"Note that Java’s URLEncoder class encodes space character(" ") into a + sign. This is contrary to other languages like Javascript that encode space character into %20."

Reply:
Just updated my explanation above, the problem is actually that URLEncoder.Encode(...) encodes spaces as '+' which is unusual for url encoding. For MQTT in particular, you can't have '+' characters in the topic string, but you can have '%20'

Reply:
So are we encoding space or "+"?

Reply:
We are fixing how ' ' gets encoded by replacing the incorrect '+' with the correct '%20'

For example:
```
String urlEncoded = URLEncoder.encode("This string has spaces", StandardCharsets.UTF_8.displayName());
String actuallyUrlEncoded = urlEncoded.replaceAll("\\+", "%20");
```

urlEncoded = "This+string+has+spaces"
actuallyUrlEncoded = "This%20string%20has%20spaces"

Reply:
C# does this correctly: https://docs.microsoft.com/en-us/dotnet/api/system.web.httputility.urlencode?view=netcore-3.1

Reply:
yes, this is what I was looking for: We are fixing how ' ' gets encoded by replacing the incorrect '+' with the correct '%20'
The initial wording of the PR confused me.

I would suggest adding this as a comment where the manual replace is done.

Reply:
Sure thing, I can add some comments to clarify this .replaceAll call